### ISSUE TITLE
Connections not properly closed on Kernel shutdown

### ISSUE THREAD

**Issue by applike-ss** [snc/SncRedisBundle#400]
I experience propblems because of too many open connections during unit tests.
These connection are at half from ongr which doesn't close the connection on kernel shutdown => Garbage collection doesn't free it.

The used PHP Version is 7.1.12
Bundle is 2.0.6

**Comment by curry684** [comment]
Could you elaborate a bit more on the specific issue you are facing?

**Comment by curry684** [comment]
If I understand correctly => if you have a ton of functional tests instantiating rebooted kernels, and therefore reinstantiating the Redis instances, they will start failing in the end indicating Redis has run out of sockets or something?

**Comment by curry684** [comment]
@applike-ss I've just typed a bit in a related report at https://github.com/snc/SncRedisBundle/pull/412#issuecomment-380919445. As said there I'm not all too excited about changing or extending perfectly fine code just to facilitate test suites if it is easier and more correct to fix the test suite instead.

What is your take on that?

**Comment by applike-ss** [comment]
@curry684 i do understand your point, but it is not the test case/suite that creates the problem. 
In my Opinion, whenever a Bundle is shut down, all connections that this bundle initiated should be properly closed instead of relying on php to close them. 
Doctrine is a good example for this behaviour. When it shuts down, it gathers all entity managers and clears their state. Afterwards it gathers all connections and closes them.
I believe, it shouldn't break anything anyway, since the connection wouldn't be used after a bundle shutdown anyway.

**Comment by applike-ss** [comment]
You are right, having lots of tests using redis, we are facing an issue with connection count increasing to the maximum of the machine running the tests.

**Comment by supersmile2009** [comment]
@applike-ss check out #415 - should be a solution you're looking for.

**Comment by applike-ss** [comment]
@supersmile2009 thanks a lot. Looking forward for it to be merged.

**Comment by curry684** [comment]
Fixed in 2.x by #415 but keeping this issue open to better investigate how to do this in 3.0 without making a service needlessly public.

**Comment by supersmile2009** [comment]
Like @dkarlovi mentioned on #415 it is possible to create env-specific service for test environment and declare it public instead of making public the original service.
However I'm not sure if there's anything else except unit tests that's using Kernel's shutdown feature. Method's docBlock says "This method is mainly useful when doing functional testing." But I'm particularly concerned about things like PPM - I haven't had a chance to try it yet, so I don't know whether it uses shutdown.

If shutdown is indeed only a unit tests thing, than a public service prefixed by `test.` and auto generated in a compiler pass is a perfect solution.
However if it's not the only case of shutdown usage, then we should probably also add a bool config option like `snc_redis.session.close_on_sutdown`. (default - false)
This way public service will be generated in test env + when this option is set to true, if someone needs it not just in test environment.

**Comment by dkarlovi** [comment]
@supersmile2009 if you wish to support PPM, the official was to do it is to tag the service with [`kernel.reset`](https://github.com/symfony/symfony/pull/24155).
